2015-11-09  Renlin Li  <renlin.li@arm.com>

	Backport from mainline
	2014-09-12  Wilco Dijkstra  <wilco.dijkstra@arm.com>

	* gcc/ree.c (combine_reaching_defs): Ensure inserted copy don't change
	the number of hard registers.

 
--- a/src/gcc/ree.c	(revision 230032)
+++ b/src/gcc/ree.c	(revision 229978)
@@ -793,14 +793,6 @@
       if (state->modified[INSN_UID (cand->insn)].kind != EXT_MODIFIED_NONE)
 	return false;
 
-      enum machine_mode dst_mode = GET_MODE (SET_DEST (PATTERN (cand->insn)));
-      rtx src_reg = get_extended_src_reg (SET_SRC (PATTERN (cand->insn)));
-
-      /* Ensure the number of hard registers of the copy match.  */
-      if (HARD_REGNO_NREGS (REGNO (src_reg), dst_mode)
-	  != HARD_REGNO_NREGS (REGNO (src_reg), GET_MODE (src_reg)))
-	return false;
-
       /* There's only one reaching def.  */
       rtx def_insn = state->defs_list[0];
 
@@ -850,7 +842,7 @@
       start_sequence ();
       rtx pat = PATTERN (cand->insn);
       rtx new_dst = gen_rtx_REG (GET_MODE (SET_DEST (pat)),
-                                 REGNO (get_extended_src_reg (SET_SRC (pat))));
+                                 REGNO (XEXP (SET_SRC (pat), 0)));
       rtx new_src = gen_rtx_REG (GET_MODE (SET_DEST (pat)),
                                  REGNO (SET_DEST (pat)));
       emit_move_insn (new_dst, new_src);
